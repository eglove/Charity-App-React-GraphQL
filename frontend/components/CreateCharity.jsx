import React, {Component} from 'react';
import {Mutation} from 'react-apollo';
import gql from "graphql-tag";
import styled from "styled-components";
import Router from "next/router";
import Head from 'next/head';
import Error from './ErrorMessage';
import Form from "./styles/Form";

const CREATE_CHARITY_MUTATION = gql`
    mutation CREATE_CHARITY_MUTATION(
        $ein: String!
        $name: String!
        $description: String!
        $website: String
        $image: String
        $largeImage: String
        $imageDescription: String
        $street: String
        $city: String
        $state: String
        $zip: String
    ) {
        createCharity(
            ein: $ein
            name: $name
            description: $description
            website: $website
            image: $image
            largeImage: $largeImage
            imageDescription: $imageDescription
            street: $street
            city: $city
            state: $state
            zip: $zip
        ) {
            id
        }
    }
`;

const Required = styled.span`
    color: red;
`;
const LookupLink = styled.a`
    text-decoration: underline;
`;

class CreateCharity extends Component {
    state = {
        ein: '',
        name: '',
        description: '',
        website: '',
        image: '',
        largeImage: '',
        imageDescription: '',
        street: '',
        city: '',
        state: '',
        zip: ''
    };

    handleChange = (e) => {
        const {name, type, value} = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({[name]: val});
    };

    uploadFile = async (e) => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        // Preset to resize images via cloudinary
        data.append('upload_preset', 'CharityApp');

        const res = await fetch(
            'https://api.cloudinary.com/v1_1/eglove/image/upload',
            {
                method: 'POST',
                body: data
            }
        );
        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url,
        })
    };

    render() {
        return (
            <Mutation mutation={CREATE_CHARITY_MUTATION} variables={this.state}>
                {(createCharity, {loading, error}) => (
                    <>
                    <Head>
                        <title>Cognitame ðŸ’™ Add Charity</title>
                    </Head>
                    <Form
                        onSubmit={async (e) => {
                            // Stop form from submitting
                            e.preventDefault();
                            // call mutation
                            const res = await createCharity();
                            // redirect to charity page
                            await Router.push({
                                pathname: '/charity',
                                query: {id: res.data.createCharity.id}
                            })
                        }}
                    >
                        <Error error={error}/>
                        <fieldset disabled={loading} aria-busy={loading}>

                            <label htmlFor="ein">
                                EIN (<LookupLink href="https://apps.irs.gov/app/eos/">Employer Identification
                                Number</LookupLink>) <Required>*</Required>
                                <input
                                    type="text"
                                    id="ein"
                                    name="ein"
                                    placeholder="EIN"
                                    value={this.state.ein}
                                    onChange={this.handleChange}
                                    required
                                />
                            </label>
                            <label htmlFor="name">
                                Name <Required>*</Required>
                                <input
                                    type="text"
                                    id="name"
                                    name="name"
                                    placeholder="Name"
                                    value={this.state.name}
                                    onChange={this.handleChange}
                                    required
                                />
                            </label>
                            <label htmlFor="description">
                                Description <Required>*</Required>
                                <textarea
                                    id="description"
                                    name="description"
                                    placeholder="Description"
                                    value={this.state.description}
                                    onChange={this.handleChange}
                                    required
                                />
                            </label>
                            <label htmlFor="website">
                                Website
                                <input
                                    type="text"
                                    id="website"
                                    name="website"
                                    placeholder="Website"
                                    value={this.state.website}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="file">
                                Image
                                <input
                                    type="file"
                                    id="file"
                                    name="file"
                                    placeholder="Upload Logo"
                                    onChange={this.uploadFile}
                                />
                                {this.state.image && (<img src={this.state.image} width="200" alt="Image Upload Preview"/>)}
                            </label>
                            <label htmlFor="imageDescription">
                                Image Description
                                <textarea
                                    id="imageDescription"
                                    name="imageDescription"
                                    placeholder="Image Description"
                                    value={this.state.imageDescription}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="street">
                                Street
                                <input
                                    type="text"
                                    id="street"
                                    name="street"
                                    placeholder="Street"
                                    value={this.state.street}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="city">
                                City
                                <input
                                    type="text"
                                    id="city"
                                    name="city"
                                    placeholder="City"
                                    value={this.state.city}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="state">
                                State
                                <input
                                    type="text"
                                    id="state"
                                    name="state"
                                    placeholder="State"
                                    value={this.state.state}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="zip">
                                Zip
                                <input
                                    type="text"
                                    id="zip"
                                    name="zip"
                                    placeholder="Zip"
                                    value={this.state.zip}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <button type="submit">Submit</button>
                        </fieldset>
                    </Form>
                    </>
                )}
            </Mutation>
        );
    }
}

export default CreateCharity;
export {CREATE_CHARITY_MUTATION};