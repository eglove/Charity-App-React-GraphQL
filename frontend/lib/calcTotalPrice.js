export default function calcTotalPrice(favorites) {
  return favorites.reduce((tally, favorite) => {
    if (!favorite.charity) return tally;
    return tally + favorite.amount;
  }, 0);
}
