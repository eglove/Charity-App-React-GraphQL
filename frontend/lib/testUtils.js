import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const fakeCharity = () => ({
  __typename: 'Charity',
  id: casual.uuid,
  ein: casual.uuid,
  name: casual.company_name,
  description: casual.description,
  website: casual.url,
  image: `${casual.word}.png`,
  largeImage: `${casual.word}.png`,
  imageDescription: casual.description,
  street: casual.street,
  city: casual.city,
  state: casual.state,
  zip: casual.zip,
  user: casual.uuid,
});

const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  password: casual.password,
  totalDonated: casual.random,
  permissions: ['ADMIN'],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeCharity,
  fakeUser,
};
