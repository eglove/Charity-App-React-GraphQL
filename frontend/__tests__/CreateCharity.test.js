import {mount} from "enzyme";
import wait from "waait";
import toJson from "enzyme-to-json";
import {MockedProvider} from "react-apollo/test-utils";
import CreateCharity, {CREATE_CHARITY_MUTATION} from "../components/CreateCharity";
import {fakeCharity} from "../lib/testUtils";

const imageLink = "https://pics.com/pic.png";
// mocked global fetch API
global.fetch = jest.fn().mockResolvedValue({
    json: () => ({
        secure_url: imageLink,
        eager: [{secure_url: imageLink}],
    })
});

describe('<CreateItem/>', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateCharity/>
            </MockedProvider>
        );

        const form = wrapper.find('form[data-test="form"]');
        expect(toJson(form)).toMatchSnapshot();
    });

    it('uploads a file when changed', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateCharity/>
            </MockedProvider>
        );

        const input = wrapper.find('input[type="file"]');
        input.simulate('change', {target: {files: ['fakeImage.png']}});
        await wait();
        const component = wrapper.find('CreateCharity').instance();
        expect(component.state.image).toEqual(imageLink);
        expect(component.state.largeImage).toEqual(imageLink);
        expect(global.fetch).toHaveBeenCalled();
        global.fetch.mockReset();
    });

    it('handles state updating', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateCharity/>
            </MockedProvider>
        );

        wrapper.find('#name').simulate('change', {target: {value: 'Testing', name: 'name'}});
        wrapper.find('#website').simulate('change', {target: {value: 'example.com', name: 'website'}});
        wrapper.find('#description').simulate('change', {target: {value: 'A description', name: 'description'}});
        wrapper.find('#city').simulate('change', {target: {value: 'Springfield', name: 'city'}});
        wrapper.find('#ein').simulate('change', {target: {value: '1234', name: 'ein'}});
        wrapper.find('#imageDescription').simulate('change', {target: {value: 'Descriptive description', name: 'imageDescription'}});
        wrapper.find('#state').simulate('change', {target: {value: 'Aknas', name: 'state'}});
        wrapper.find('#street').simulate('change', {target: {value: '123 Main', name: 'street'}});
        wrapper.find('#zip').simulate('change', {target: {value: '00001', name: 'zip'}});

        expect(wrapper.find('CreateCharity').instance().state).toMatchObject({
            name: 'Testing',
            website: 'example.com',
            description: 'A description',
            city: "Springfield",
            ein: "1234",
            image: "",
            imageDescription: "Descriptive description",
            largeImage: "",
            state: "Aknas",
            street: "123 Main",
            zip: "00001",
        });
    });

    xit('create a charity when form is submitted', async () => {
        const charity = fakeCharity();
        const mocks = [
            {
                request: {
                    query: CREATE_CHARITY_MUTATION,
                    variables: {
                        name: charity.name,
                        website: charity.website,
                        description: charity.description,
                        city: charity.city,
                        ein: charity.ein,
                        image: charity.image,
                        imageDescription: charity.imageDescription,
                        largeImage: charity.largeImage,
                        state: charity.state,
                        street: charity.street,
                        zip: charity.zip,
                    },
                },

                result: {
                    data: {
                        createCharity: {
                            ...fakeCharity,
                            typeName: 'Charity',
                        },
                    },
                },
            },
        ];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <CreateCharity/>
            </MockedProvider>
        );

        // simulate someone filling out form
        wrapper.find('#name').simulate('change', {target: {value: charity.name, name: 'name'}});
        wrapper.find('#website').simulate('change', {target: {value: charity.website, name: 'website'}});
        wrapper.find('#description').simulate('change', {target: {value: charity.description, name: 'description'}});
        wrapper.find('#city').simulate('change', {target: {value: charity.city, name: 'city'}});
        wrapper.find('#ein').simulate('change', {target: {value: charity.ein, name: 'ein'}});
        wrapper.find('#file').simulate('change', {target: {value: "", name: 'file'}});
        // wrapper.find('#largeImage').simulate('change', {target: {value: "", name: 'largeImage'}});
        wrapper.find('#imageDescription').simulate('change', {target: {value: charity.imageDescription, name: 'imageDescription'}});
        wrapper.find('#state').simulate('change', {target: {value: charity.state, name: 'state'}});
        wrapper.find('#street').simulate('change', {target: {value: charity.street, name: 'street'}});
        wrapper.find('#zip').simulate('change', {target: {value: charity.zip, name: 'zip'}});

        // mock the router
        Router.router = {push: jest.fn() };
        wrapper.find('form').simulate('submit');
        await wait(50);
        expect(Router.router.push).toHaveBeenCalled();
    });
});
