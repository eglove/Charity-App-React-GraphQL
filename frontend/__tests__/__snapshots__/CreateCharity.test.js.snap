// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> renders and matches snapshot 1`] = `
<form
  className="Form-feif0j-0 hNIufj"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="ein"
    >
      EIN (
      <CreateCharity__LookupLink
        href="https://apps.irs.gov/app/eos/"
      >
        <a
          className="CreateCharity__LookupLink-sc-7bw0bv-1 hWuqKe"
          href="https://apps.irs.gov/app/eos/"
        >
          Employer Identification Number
        </a>
      </CreateCharity__LookupLink>
      ) 
      <CreateCharity__Required>
        <span
          className="CreateCharity__Required-sc-7bw0bv-0 eHGSTg"
        >
          *
        </span>
      </CreateCharity__Required>
      <input
        id="ein"
        name="ein"
        onChange={[Function]}
        placeholder="EIN"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name 
      <CreateCharity__Required>
        <span
          className="CreateCharity__Required-sc-7bw0bv-0 eHGSTg"
        >
          *
        </span>
      </CreateCharity__Required>
      <input
        id="name"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="description"
    >
      Description 
      <CreateCharity__Required>
        <span
          className="CreateCharity__Required-sc-7bw0bv-0 eHGSTg"
        >
          *
        </span>
      </CreateCharity__Required>
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        value=""
      />
    </label>
    <label
      htmlFor="website"
    >
      Website
      <input
        id="website"
        name="website"
        onChange={[Function]}
        placeholder="Website"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload Logo"
        type="file"
      />
    </label>
    <label
      htmlFor="imageDescription"
    >
      Image Description
      <textarea
        id="imageDescription"
        name="imageDescription"
        onChange={[Function]}
        placeholder="Image Description"
        value=""
      />
    </label>
    <label
      htmlFor="street"
    >
      Street
      <input
        id="street"
        name="street"
        onChange={[Function]}
        placeholder="Street"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="city"
    >
      City
      <input
        id="city"
        name="city"
        onChange={[Function]}
        placeholder="City"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="state"
    >
      State
      <input
        id="state"
        name="state"
        onChange={[Function]}
        placeholder="State"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="zip"
    >
      Zip
      <input
        id="zip"
        name="zip"
        onChange={[Function]}
        placeholder="Zip"
        type="text"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
